### PR TITLE
[RFC 0043] RFC Steering Committee Rotation

### PR DESCRIPTION
**Pull request by globin**
This is a draft RFC to implement rotation for the RFCSC.

cc @grahamc @NixOS/rfc-steering-committee @ciil

[Rendered Version](https://github.com/mayflower/nixos-rfcs/blob/rfcsc-rotation/rfcs/0043-rfcsc-rotation.md)

### CODE REVIEW

**Comment by globin** [comment]
I nominate @grahamc @zimbatm @edolstra and @flokli for shepherding.

File: rfcs/0043-rfcsc-rotation.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+complexity of having to hold elections, including defining who is eligible to
+vote and how to hold them.
+
+# Alternatives
+[alternatives]: #alternatives
+
+Do nothing: then the current members of the RFC Steering Committee as defined
+in RFC 36 could stay on the Committee indefinitely or at least until that part
+of RFC 36 is overridden by a newly accepted RFC.
+
```
Comment (7c6f434c) [inline rfcs/0043-rfcsc-rotation.md]: I think «rigorously defining voting eligibility and an election procedure» is a plausible alternative, even if a costly one, and should be listed here.

**Comment by Ericson2314** [comment]
This is similar to how release managers rotate, and that's process has been a wonderful success, so +1 from me. \[That isn't to say -1 to other systems, however. I'm just approving of this one.\]

**Comment by 7c6f434c** [comment]
@domenkozar as a non-native speaker with a large enough passive vocabulary I didn't pay attention to the word choice; now I looked up in a dictionary and maybe people who remember the _exact_ definition would _also_ think the text with «resignation» is easier to read (abdication implies some position held for life).

**Comment by globin** [comment]
I have addressed the comments above.

**Comment by 7c6f434c** [comment]
I guess we should nominate some more people as potential Shepherds, because it would be a bit weird to have an all-Tweag Shepherd team on this specific RFC…

**Comment by zimbatm** [comment]
good call @7c6f434c :)

**Comment by globin** [comment]
Also nominating @lheckemann, @7c6f434c, @ryantm and @domenkozar.

Yeah, will add this to the README similar as #38 was to #36!

**Comment by 7c6f434c** [comment]
Re: nomination — thanks, I accept. But without promises that scheduling a sycnhronous chat will be easy. (But maybe reasonably fast responses to email will be enough anyway, and that I do promise).

File: rfcs/0043-rfcsc-rotation.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+themself or any other person who accepts the nomination. Members of the
+previous RFCSC explicitly can stand again, but should reflect on their free
+time and commitment to their role. The nomination period starts at the
+beginning of November, a minimum of four weeks before election, and is
+announced on all relevant communication channels (as of April 2019,
+discourse.nixos.org, IRC #nixos and #nixos-dev, and NixOS Weekly).
+
+Additionally to RFC 36 a new restriction formally comes into effect. In order
+to avoid conflict of interest there is an upper cap of 2 members working for a
+single employer.
```
Comment (7c6f434c) [inline rfcs/0043-rfcsc-rotation.md]: 1. Does that mean that if a member changes employment in a cap-violating way, they are expected to resign (unless one of the two members from the new employment place prefers to resign themselves)? Should this be explicitly recorded? 

2. I want to have this question on the record: do we want to respect the other dimensions of conflict of interest? I think it would be feasible to ask that if a full SC of five members is being selected, at least one member of the incoming committee should have used Nix outside the single most popular setup (currently NixOS on `x86_64`). It would be a nightmare in a large election procedure, but for unanimous selection of successors it seems feasible.

File: rfcs/0043-rfcsc-rotation.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+The members chosen through the original implementation as
+specified in RFC 36 are regular members as specified in this RFC. They will
+stay on as members until replaced by new members as outlined below.
+
+If a seat has to be filled earlier than at the yearly vote, the new member will
+only serve for the rest of the term.
+
+Each year at the first meeting of the RFCSC in December (starting 2019,
+approximately a year after establishment in RFC 36) they unanimously decide on
+the succeeding committee members. If unanimous agreement cannot be reached, the
```
Comment (7c6f434c) [inline rfcs/0043-rfcsc-rotation.md]: Does ths old SC hold one more meeting (so that the new SC has a week to agree on their meeting schedule without interruption in RFC processing)? Or does the old SC continue to work until Christmas-New Year week (when everybody expects interruptions anyway)?

File: rfcs/0043-rfcsc-rotation.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+With the implementation of [RFC
+36](https://github.com/NixOS/rfcs/blob/master/rfcs/0036-rfc-process-team-amendment.md)
+the RFC Steering Committee has been established. Among others, future work for
+that RFC included a definition of how members to the Committee are to be chosen
+or removed. This RFC provides mechanisms through which membership in the
+Steering Committee can be established or ended, so as not to make membership a
+mandate for life, allowing a healthy rotation of members, and offering several
+procedures to always keep the Steering Committee capable of making decisions
+such that the RFC process can keep running smoothly at all times.
```
Comment (lheckemann) [inline rfcs/0043-rfcsc-rotation.md]: Run-on sentence! :rotating_light: The "mandate for life" and "healthy rotation" bits are also just restating that the committee will change and that that is in some vague way good. Here's an attempt at rephrasing with a bit more structure and justification:

> This RFC provides mechanisms for beginning and ending Steering Committee membership. The purpose of this is to ensure that the committee:
> - continues to consist of well-informed and motivated members;
> - continues to represent the evolving community;
> - remains active and keeps the RFC process running.

**Comment by ryantm** [comment]
I accept @globin 's nomination.

File: rfcs/0043-rfcsc-rotation.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+The RFC Steering Committee shall always have five members. If membership drops
+below five members (for example by resignation of a member as detailed below), a
+new member shall directly be elected after a nomination period of at least two
+weeks (see below for nomination and selection process). If the number of
+members of the RFCSC drops below 4 people, it cannot proceed with shepherd team
+selections until new members have been selected.
+
+## Ending membership
+A member can end their membership in the Steering
+Committee by either of the following four mechanisms, ordered from most
```
[the remark this answers is not in the document]
Comment (7c6f434c) [reply rfcs/0043-rfcsc-rotation.md]: I guess the order is intended to vaguely guide the norms and expectations without creating formal rules.

File: rfcs/0043-rfcsc-rotation.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+Each RFC Steering Committee (RFCSC) unanimously elects the succeeding one at
+their first meeting in December from an open list of nominees.
+
+# Motivation
+[motivation]: #motivation
+
+With the implementation of [RFC
+36](https://github.com/NixOS/rfcs/blob/master/rfcs/0036-rfc-process-team-amendment.md)
+the RFC Steering Committee has been established. Among others, future work for
```
[the remark this answers is not in the document]
Comment (ciil) [reply rfcs/0043-rfcsc-rotation.md]: That's debatable, at least from a common law framing where the set phrase "included but not limited to" sometimes does kind of imply the completeness of includes (and is an instance of the legal principle _expressio unius est exclusio alterius_). That is, of course, just an equally debatable nitpick in the other direction :smile:. I do think we can simply remove the `among others` without loss of meaning. We're not negotiating a liability waiver, after all. For me personally however, it adds to the flow just enough to make it worthwhile.

File: rfcs/0043-rfcsc-rotation.md [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+mandate for life, allowing a healthy rotation of members, and offering several
+procedures to always keep the Steering Committee capable of making decisions
+such that the RFC process can keep running smoothly at all times.
+
+# Detailed design
+[design]: #detailed-design
+
+The RFC Steering Committee shall always have five members. If membership drops
+below five members (for example by resignation of a member as detailed below), a
+new member shall directly be elected after a nomination period of at least two
```
[the remark this answers is not in the document]
Comment (ciil) [reply rfcs/0043-rfcsc-rotation.md]: Maybe `shall be elected without delay` would make this clearer?

**Comment by edolstra** [comment]
@domenkozar also volunteered. So we have the following shepherds: @ryantm , @lheckemann, @7c6f434c and @domenkozar. Thanks!

@lheckemann Do you want to lead the shepherd team on this one?

**Comment by lheckemann** [comment]
I can, but it may take a bit longer if I do since I've got a fair bit on my plate in the next few weeks. If everyone's fine with that, :+1:

File: rfcs/0043-rfcsc-rotation.md (hunk 2 above)
Comment (globin) [reply rfcs/0043-rfcsc-rotation.md]: 1. I don't think it is necessary, but if someone cares to resign no one will stop them.
2. I also think we don't have to specify this, as the RFCSC should always try and set up a balanced shepherding team, familiar with the affected parts and the RFCSC can realistically not cover everything themself.

Comment (7c6f434c) [reply rfcs/0043-rfcsc-rotation.md]: 1. I am OK with that, but then maybe reword it to say that there is a cap of _appointing_ at most 2 people working for the same employer? Because right now it seems to mean my interpretation.

2. It might be easier to evaluate experience (and relevance of that experience) of people working on special use cases for a SC including someone who doesn't use the «default setup» exclusively. But I do not insist.

**Comment by globin** [comment]
I've tried to address all reviews, feel free to nitpick further! :hammer_and_pick:

**Comment by 7c6f434c** [comment]
I have gone through the discussion, there is one nit not addressed in any way (except a :+1: from an RFC coauthor):

> Rules nit: what if two members are absent?

(re: unanimous vote for removing the person due to missing the meetings)

@globin if you think that this situation is not worth an explicit mention, please say so.

File: rfcs/0043-rfcsc-rotation.md [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+[unresolved]: #unresolved-questions
+
+As of now, none.
+
+# Future work
+[future]: #future-work
+
+As this process is to be implemented over a fairly long time frame (a year for
+each iteration), this framework might have to be revised at a later date,
+incorporating any learnings made over these years.
```
[the remark this answers is not in the document]
Comment (domenkozar) [reply rfcs/0043-rfcsc-rotation.md]: @leckermann btw you can click the "plus minus" button when commenting to suggest a wording change like this :)

**Comment by globin** [comment]
I had missed that, but we have just discussed it and came to the conclusion that we would send out an extra invitation to the meeting to all members where there would be a vote to remove a member for the team. All members attending the meeting will then vote on removing the member, objections by members who cannot attend can obviously object prior to the meeting.

I personally think we do not have to specify this in the RFC because all of this relies on good faith anyway but if someone prefers that, I can do so.

Are there further remarks, objections, suggestions etc.?

**Comment by ciil** [comment]
_If_ so desired, my suggestion for phrasing would be something like `The meeting to remove a member from the RFCSC shall constitute a quorum without consideration for the number of members present.` as that doesn't depend on any specific number of people present.

But for now, I'd rather go the good faith route as well.

**Comment by lheckemann** [comment]
>The meeting to remove a member from the RFCSC shall constitute a quorum without consideration for the number of members present.

So could a quorum consisting entirely of non-members remove members? :thinking:

Yeah, I think good faith makes sense here :D

**Comment by 7c6f434c** [comment]
@lheckemann if there was no agreement in RFCSC on moving the meeting and the regularly scheduled meeting is only attended by non-members, maybe we can count that as «no objections»…

I agree with the interpretation that «agreement of all the other members» means «lack of objections, with objections accepted both by email and at the meeting itself», so there is probably no need to change the RFC.

I still wanted every point raised to be explicitly acknowledged, thanks to everyone for your cooperation.

File: rfcs/0043-rfcsc-rotation.md [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+The RFC Steering Committee shall always have five members. If membership drops
+below five members (for example by resignation of a member as detailed below),
+a new member shall be elected without delay after a nomination period of at
+least two weeks (see below for nomination and selection process). If the number
+of members of the RFCSC drops below 4 people, it cannot proceed with shepherd
+team selections until new members have been selected.
+
+## Ending membership
+A member can end their membership in the Steering
+Committee by either of the following four mechanisms:
```
Comment (7c6f434c) [inline rfcs/0043-rfcsc-rotation.md]: Maybe rephrase as «Membership of a member in the Steering Committee can end by …»

(I _tried_ to do a two-line suggestion and failed)

**Comment by Mic92** [comment]
We decided on @lheckemann as leader

**Comment by 7c6f434c** [comment]
@globin what's your opinion on rephrasing?

**Comment by globin** [comment]
Updated

**Comment by 7c6f434c** [comment]
@globin Thanks

@lheckemann @ryantm @domenkozar I now see no outstanding questions (even minor ones), and the last new comment related to text was more than ten days ago. As far as I see, nobody has objected to the RFC in general. I think it is time to vote on starting the FCP with disposition to accept. For the record: I am in favour.

**Comment by ryantm** [comment]
I sign off on @7c6f434c 's motion to enter the FCP with disposition to merge.

**Comment by ryantm** [comment]
@lheckemann @domenkozar Can you indicate your sign off or objection to entering FCP?

**Comment by domenkozar** [comment]
@ryantm I'll have time to reread the RFC tomorrow and sign off.

File: rfcs/0043-rfcsc-rotation.md [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+## Ending membership
+A member can end their membership in the Steering
+Committee by either of the following four mechanisms, ordered from most
+frequent to least:
+
+1. At the end of an election period.
+2. Resignation
+3. Unanimous vote by all other members after having missed two or more regular
+   meetings without giving an appropriate excuse.
```
[the remark this answers is not in the document]
Comment (domenkozar) [reply rfcs/0043-rfcsc-rotation.md]: Doesn't have to be on the same meeting, vote would be cast per member.

File: rfcs/0043-rfcsc-rotation.md (hunk 3 above)
Comment (domenkozar) [reply rfcs/0043-rfcsc-rotation.md]: It takes immediate effect at the beginning of December.

The motivation is to not tie things around the new year when there's already lots of personal errands to be completed :)

Comment (7c6f434c) [reply rfcs/0043-rfcsc-rotation.md]: Immediate takeover has the problem of coordination; from the point of view of end-of-year, second half of December and first half of January are all bad anyway.

File: rfcs/0043-rfcsc-rotation.md [hunk 11]
```diff
@@ -36,8 +36,8 @@ of members of the RFCSC drops below 4 people, it cannot proceed with shepherd
 team selections until new members have been selected.
 
 ## Ending membership
-A member can end their membership in the Steering
-Committee by either of the following four mechanisms:
+Membership in the Steering Committee can end by either of the following four
```
Comment (lheckemann) [inline rfcs/0043-rfcsc-rotation.md]: nit: s/either/any/

**Comment by ryantm** [comment]
@NixOS/rfc-steering-committee Who is in charge of making sure an FCP is "advertised widely"?

**Comment by shlevy** [comment]
@ryantm Shepherd team/leader.

File: rfcs/0043-rfcsc-rotation.md (hunk 3 above)
Comment (7c6f434c) [reply rfcs/0043-rfcsc-rotation.md]: (Just in case: the literal current text of RFC explicitly sets January as the takeover time)

File: rfcs/0043-rfcsc-rotation.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (vcunat) [reply rfcs/0043-rfcsc-rotation.md]: Do you mean to add that
> If the number of members of the RFCSC drops below 4 people

it also shouldn't be able to apply (3) or (4)?

If it was about members being _present_, I thought it's clear that "all other" means all other _on the committee_ (not just present ATM).

[the remark this answers is not in the document]
Comment (7c6f434c) [reply rfcs/0043-rfcsc-rotation.md]: I think the natural interpretation here is «all the other current members», I think reaching 3 instead of 2 is unfortunate and reaching 4 is outright bad whatever you do, and I think in both cases it is less bad to thave the possibility of removal before selecting replacement.

**Comment by lheckemann** [comment]
The FCP has passed without any objections :tada: If anyone with the appropriate permissions could merge the RFC and fix "either" → "any", I would be very grateful :)
